runtime/compiler/trj9/optimizer/ValueJProfiling.cpp

```diff
@@ -0,0 +1,883 @@
+ *
+ * \param value Node representing the value to add to the table.
+ * \param table Address of the table's base.
+ */
+TR::Node *
+TR_ValueJProfiling::createHelperCall(TR::Compilation *comp, TR::Node *value, TR::Node *table)
+   {
+   TR::SymbolReference *profiler;
+   if (value->getDataType() == TR::Int32)
+      profiler = comp->getSymRefTab()->findOrCreateRuntimeHelper(TR_jProfile32BitValue, false, false, true);
+   else
+      profiler = comp->getSymRefTab()->findOrCreateRuntimeHelper(TR_jProfile64BitValue, false, false, true);
+
+#if defined(TR_HOST_POWER) || defined(TR_HOST_ARM)
+   profiler->getSymbol()->castToMethodSymbol()->setLinkage(TR_System);
+#else
+#ifndef TR_HOST_X86
+   profiler->getSymbol()->castToMethodSymbol()->setPreservesAllRegisters();
+#endif
+   profiler->getSymbol()->castToMethodSymbol()->setSystemLinkageDispatch();
+#endif
+
+   TR::Node *helperCall = TR::Node::createWithSymRef(value, TR::call, 2, profiler);
+   helperCall->setAndIncChild(0, value);
+   helperCall->setAndIncChild(1, table);
+   return TR::Node::create(TR::treetop, 1, helperCall);
+   }
+
+/*
+ * Increment a memory address by 1. Used to increment the table's counters on both the match and other
+ * paths.
+ *
+ * \param counterType The memory addresses type, usually Int32 or Int64.
+ * \param address The memory address to increment.
+ */
+TR::Node *
+TR_ValueJProfiling::incrementMemory(TR::Compilation *comp, TR::DataType counterType, TR::Node *address)
+   {
+   TR::SymbolReference *symRef = comp->getSymRefTab()->findOrCreateArrayShadowSymbolRef(counterType, NULL);
+
+   TR::Node *load = TR::Node::createWithSymRef(address, indirectLoad(counterType), 1, address, symRef);
+   TR::Node *incCounter = TR::Node::create(address, counterType == TR::Int32 ? TR::iadd : TR::ladd, 2, load,
+      TR::Node::create(address, loadConst(counterType), 0, 1));
+   TR::Node *storeCounter = TR::Node::createWithSymRef(address, indirectStore(counterType), 2,
+      address, symRef);
+   storeCounter->setAndIncChild(1, incCounter);
+
+   return storeCounter;
+   }
+
+/*
+ * Generate a constant node matching the system's address width. Used for address calculations.
+ *
+ * \param example Example node to copy BCI.
+ * \param value Value to store in const.
+ */
+TR::Node *
+TR_ValueJProfiling::systemConst(TR::Node *example, uint64_t value)
+   {
+   TR::ILOpCodes constOp = TR::Compiler->target.is64Bit() ? TR::lconst : TR::iconst;
+   return TR::Node::create(example, constOp, 0, value);
+   }
+
+/*
+ * Its sometimes necessary to convert values to their rounded integer representation, in Int32
+ * or Int64, or to convert values to the system type, again in Int32 or Int64. This method
+ * handles those conversions as necessary. All conversions are unsigned.
+ *
+ * \param index The value to convert.
+ * \param dataType The datatype to convert to.
+ */
+TR::Node *
+TR_ValueJProfiling::convertType(TR::Node *index, TR::DataType dataType)
```

Comment:
There isn't an OMR helper for this that could be used?

Reply:
Yes, `getProperConversion` has a unsigned option I wasn't aware of. This has been corrected.